sklearn/cross_validation.py

```diff
@@ -332,6 +332,119 @@ def __len__(self):
         return self.n_folds
 
 
+def disjoint_label_folds(y, n_folds=3):
+    """Creates folds where a same label is not in two different folds.
+    
+    Parameters
+    ----------
+    y: numpy array, shape (n_samples,)
+        Contains an id for each sample.
+        The folds are built so that the same id doesn't appear in two different folds.
+    
+    n_folds: int, default=3
+        Number of folds to split the data into.
+        
+    Returns
+    -------
+    folds: numpy array of shape (n_samples, )
+        Array of integers between 0 and (n_folds - 1).
+        Folds[i] contains the folds to which sample i is assigned.
+        
+    Notes
+    -----
+    The folds are built by distributing the labels by frequency of appearance.
+    """
+    labels = np.array(y)
+    unique_labels, y = np.unique(labels, return_inverse=True)
+    
+    # number of occurrence of each label (its "weight")
+    samples_per_label = np.bincount(y)
+    # We want to distribute the most frequent labels first
+    ind = np.argsort(samples_per_label)[::-1]
```

Comment:
I wonder if we should not use a stable stort algorithm like `np.argsort(samples_per_label, kind="mergesort")` here.

@larsmans any idea if a non-stable sort could cause reproducibility issues in this case?